File: src-java/recursion/PrintList.java
package recursion;

import acm.program.ConsoleProgram;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;

public class PrintList extends ConsoleProgram {

    void printList(Queue<Integer> queue) {
        if (queue.isEmpty())   // base case
            return;

        int i = queue.poll();
        print(i + ", ");   // insertion order
        printList((queue));
        print(i + ", ");   // reverse order
    }

    @Override
    public void run() {
        setFont("Arial-bold-32");

        Queue<Integer> queue = new LinkedList<>();

        queue.add(4);
        queue.add(9);
        queue.add(2);
        queue.add(10);
        queue.add(34);
        queue.add(76);

        printList(queue);
    }
}

File: src-python/5-Recursion/PrintList.py
from collections import deque

def print_list(queue: deque[int]):
    if len(queue) <= 0:  # base case
        return

    i = queue.popleft()
    print(f"{i}, ", end='')  # insertion order
    print_list(queue)  # recursive case
    print(f"{i}, ", end='')  # reverse order

def main():
    queue = deque([4, 9, 2, 10, 34, 76])
    print_list(queue)

if __name__ == "__main__":
    main()